lib/transaction.py

```diff
@@ -689,20 +689,48 @@ def nHashType(cls):
         warnings.warn("warning: deprecated tx.nHashType()", FutureWarning, stacklevel=2)
         return 0x01 | (cls.SIGHASH_FORKID + (cls.FORKID << 8))
 
-    def serialize_preimage(self, i, nHashType=0x00000041):
+    def calc_common_sighash(self, use_cache=False):
+        """ Calculate the common sighash components that are used by
+        transaction signatures. If `use_cache` enabled then this will return
+        already-computed values from the `._cached_sighash` attribute, or
+        compute them if necessary (and then store).
+
+        For transactions with N inputs and M outputs, calculating all sighashes
+        takes only O(N + M) with the cache, as opposed to O(N^2 + NM) without
+        the cache.
+
+        Returns three 32-long bytes objects: (hashPrevouts, hashSequence, hashOutputs).
+
+        Warning: If you modify non-signature parts of the transaction
+        afterwards, this cache will be wrong! """
+        try:
+            if use_cache:
+                return self._cached_sighash
+        except AttributeError:
+            pass
+
+        inputs = self.inputs()
+        outputs = self.outputs()
+        hashPrevouts = Hash(bfh(''.join(self.serialize_outpoint(txin) for txin in inputs)))
+        hashSequence = Hash(bfh(''.join(int_to_hex(txin.get('sequence', 0xffffffff - 1), 4) for txin in inputs)))
+        hashOutputs = Hash(bfh(''.join(self.serialize_output(o) for o in outputs)))
+
+        res = hashPrevouts, hashSequence, hashOutputs
+        if use_cache:
```

Comment:
Would you mind terribly if I asked you if we were to unconditonally overwrite the cache here?

I can't imagine a benefit for not at least deleting the old value from the cache once a new value is computed.  It seems like it's possible to do the following:

1. Compute pre-image with `use_cache=True`.
2. Change your ming and modify inputs() or whatever
3. Compute pre-image on first input with `use_cache = False`
4. Compute pre-image on next input with `use_cache = True`  <--- here you get the "old" cache value from (1) above.

There's no encapsulated way to clobber the cache short of `del self._cached_sighash`....

I propose that if we unconditionally write to that cache each time a result is computed -- then at least there is some (esoteric) usage pattern (namely `use_cache=False` for first input, and `use_cache=True` for subsequent inputs) to at least kill the cache.

Just thoughts... your thoughtS?